### ISSUE TITLE
[Form Container] Proxy extension model class of Form Container results in infinite loop

### ISSUE THREAD

**Issue by HitmanInWis** [adobe/aem-core-wcm-components#239]
## Bug Report

**Current Behavior**
Extending the Container interface and implementing a new model class that proxies the WCM Core ContainerImpl results in an infinite loop on sling model instantiation for a Form Container added to a page.  The issue is coming from the `@PostConstruct` implementation of ContainerImpl, which is calling `FormsHelper` which eventually executes a sling dispatcher include that attempts to adapt to the sling model again, and thus the cycle starts.

**Expected behavior/code**
Extension via the proxy method is expected to be supported, to allow functionality to be added/modified from provided defaults.

**Attempted implementation**
```
import aQute.bnd.annotation.ConsumerType;
import com.adobe.cq.wcm.core.components.models.form.Container;

@ConsumerType
public interface HS2Container extends Container {

}
```

```
import com.adobe.cq.export.json.ComponentExporter;
import com.adobe.cq.wcm.core.components.models.form.Container;
import com.hs2solutions.aem.base.core.models.form.HS2Container;
import java.util.Map;
import org.apache.sling.api.SlingHttpServletRequest;
import org.apache.sling.api.resource.Resource;
import org.apache.sling.models.annotations.Model;
import org.apache.sling.models.annotations.Via;
import org.apache.sling.models.annotations.injectorspecific.Self;
import org.apache.sling.models.annotations.via.ForcedResourceType;

@Model(
        adaptables = {Resource.class, SlingHttpServletRequest.class},
        adapters = {Container.class, HS2Container.class},
        resourceType = "hs2/components/content/form/container"
)
public class HS2ContainerImpl implements HS2Container {
    @Self
    @Via(type = ForcedResourceType.class, value = "core/wcm/components/form/container/v2/container")
    private Container superTypeContainer;

    @Override
    public String getMethod() {
        return superTypeContainer.getMethod();
    }

    @Override
    public String getAction() {
        return superTypeContainer.getAction();
    }

    @Override
    public String getId() {
        return superTypeContainer.getId();
    }

    @Override
    public String getName() {
        return superTypeContainer.getName();
    }

    @Override
    public String getEnctype() {
        return superTypeContainer.getEnctype();
    }

    @Override
    public String getResourceTypeForDropArea() {
        return superTypeContainer.getResourceTypeForDropArea();
    }

    @Override
    public String getRedirect() {
        return superTypeContainer.getRedirect();
    }

    @Override
    public String[] getExportedItemsOrder() {
        return superTypeContainer.getExportedItemsOrder();
    }

    @Override
    public Map<String, ? extends ComponentExporter> getExportedItems() {
        return superTypeContainer.getExportedItems();
    }

    @Override
    public String getExportedType() {
        return superTypeContainer.getExportedType();
    }
}
```

**Resulting stack trace loop**
```
	at org.apache.sling.models.impl.ModelAdapterFactory.invokePostConstruct(ModelAdapterFactory.java:896)
	at org.apache.sling.models.impl.ModelAdapterFactory.createObject(ModelAdapterFactory.java:703)
	at org.apache.sling.models.impl.ModelAdapterFactory.internalCreateModel(ModelAdapterFactory.java:394)
	at org.apache.sling.models.impl.ModelAdapterFactory.createModel(ModelAdapterFactory.java:261)
	at org.apache.sling.scripting.sightly.models.impl.SlingModelsUseProvider.provide(SlingModelsUseProvider.java:126)
	at org.apache.sling.scripting.sightly.impl.engine.extension.use.UseRuntimeExtension.call(UseRuntimeExtension.java:73)
	at org.apache.sling.scripting.sightly.impl.engine.runtime.RenderContextImpl.call(RenderContextImpl.java:69)
	at org.apache.sling.scripting.sightly.apps.core.wcm.components.form.container.v2.container.container_html.render(container_html.java:47)
	at org.apache.sling.scripting.sightly.java.compiler.RenderUnit.render(RenderUnit.java:48)
	at org.apache.sling.scripting.sightly.impl.engine.SightlyCompiledScript.eval(SightlyCompiledScript.java:61)
	at org.apache.sling.scripting.core.impl.DefaultSlingScript.call(DefaultSlingScript.java:386)
	at org.apache.sling.scripting.core.impl.DefaultSlingScript.eval(DefaultSlingScript.java:184)
	at org.apache.sling.scripting.core.impl.DefaultSlingScript.service(DefaultSlingScript.java:491)
	at org.apache.sling.engine.impl.request.RequestData.service(RequestData.java:552)
	at org.apache.sling.engine.impl.filter.SlingComponentFilterChain.render(SlingComponentFilterChain.java:44)
	at org.apache.sling.engine.impl.filter.AbstractSlingFilterChain.doFilter(AbstractSlingFilterChain.java:77)
	at com.day.cq.wcm.core.impl.WCMDeveloperModeFilter.doFilterWithErrorHandling(WCMDeveloperModeFilter.java:164)
	at com.day.cq.wcm.core.impl.WCMDeveloperModeFilter.doFilter(WCMDeveloperModeFilter.java:135)
	at org.apache.sling.engine.impl.filter.AbstractSlingFilterChain.doFilter(AbstractSlingFilterChain.java:68)
	at com.day.cq.wcm.core.impl.WCMDebugFilter.doFilter(WCMDebugFilter.java:138)
	at org.apache.sling.engine.impl.filter.AbstractSlingFilterChain.doFilter(AbstractSlingFilterChain.java:68)
	at com.day.cq.wcm.core.impl.WCMComponentFilter.doFilter(WCMComponentFilter.java:165)
	at org.apache.sling.engine.impl.filter.AbstractSlingFilterChain.doFilter(AbstractSlingFilterChain.java:68)
	at com.day.cq.wcm.core.impl.PageLockFilter.doFilter(PageLockFilter.java:91)
	at org.apache.sling.engine.impl.filter.AbstractSlingFilterChain.doFilter(AbstractSlingFilterChain.java:68)
	at com.day.cq.personalization.impl.TargetComponentFilter.doFilter(TargetComponentFilter.java:96)
	at org.apache.sling.engine.impl.filter.AbstractSlingFilterChain.doFilter(AbstractSlingFilterChain.java:68)
	at org.apache.sling.engine.impl.SlingRequestProcessorImpl.processComponent(SlingRequestProcessorImpl.java:282)
	at org.apache.sling.engine.impl.SlingRequestProcessorImpl.dispatchRequest(SlingRequestProcessorImpl.java:322)
	at org.apache.sling.engine.impl.request.SlingRequestDispatcher.dispatch(SlingRequestDispatcher.java:211)
	at org.apache.sling.engine.impl.request.SlingRequestDispatcher.include(SlingRequestDispatcher.java:104)
	at com.day.cq.wcm.core.impl.WCMComponentFilter$ForwardRequestDispatcher.include(WCMComponentFilter.java:516)
	at com.day.cq.wcm.foundation.forms.FormsHelper.includeResource(FormsHelper.java:598)
	at com.day.cq.wcm.foundation.forms.FormsHelper.runAction(FormsHelper.java:267)
	at com.adobe.cq.wcm.core.components.internal.models.v1.form.ContainerImpl.runActionTypeInit(ContainerImpl.java:146)
	at com.adobe.cq.wcm.core.components.internal.models.v1.form.ContainerImpl.initModel(ContainerImpl.java:136)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.sling.models.impl.ModelAdapterFactory.invokePostConstruct(ModelAdapterFactory.java:896)
	at org.apache.sling.models.impl.ModelAdapterFactory.createObject(ModelAdapterFactory.java:703)
	at org.apache.sling.models.impl.ModelAdapterFactory.internalCreateModel(ModelAdapterFactory.java:394)
	at org.apache.sling.models.impl.ModelAdapterFactory.createModel(ModelAdapterFactory.java:261)
	at org.apache.sling.scripting.sightly.models.impl.SlingModelsUseProvider.provide(SlingModelsUseProvider.java:126)
	at org.apache.sling.scripting.sightly.impl.engine.extension.use.UseRuntimeExtension.call(UseRuntimeExtension.java:73)
	at org.apache.sling.scripting.sightly.impl.engine.runtime.RenderContextImpl.call(RenderContextImpl.java:69)
	at org.apache.sling.scripting.sightly.apps.core.wcm.components.form.container.v2.container.container_html.render(container_html.java:47)
```

**Comment by jckautzmann** [comment]
@HitmanInWis Thx for reporting this issue. We've added it to our backlog (CQ-4248351).

**Comment by jckautzmann** [comment]
hi @HitmanInWis , please have a look at our recommendations regarding the delegation pattern:
https://github.com/adobe/aem-core-wcm-components/wiki/Delegation-Pattern-for-Sling-Models
They are the best practices that we recommend to follow.

Back to your code: The first attempt to use the delegation pattern with @Via seems ok as Sling will try to adapt the Resource to the Image model:
```
    @Self
    @Via(type = ForcedResourceType.class, value = "core/wcm/components/image/v2/image")
    private Image superTypeImage;
```
The 2nd attempt seems wrong as according to the Sling docs, @ChildResource will inject a child resource by name, not a model:
    @ChildResource
    private HS2Image image;
```
I would suggest that you refactor your code by following our delegation pattern guidelines.

**Comment by jckautzmann** [comment]
I'm closing this issue for now. Please reopen if needed.

**Comment by jckautzmann** [comment]
Please follow our guidelines around delegation pattern:
https://github.com/adobe/aem-core-wcm-components/wiki/Delegation-Pattern-for-Sling-Models

Closing the issue for now. If the issue is still there, please reopen it.

**Comment by HitmanInWis** [comment]
@jckautzmann I believe my solution above is already using the pattern delineated on the linked page.

Can you help me understand what I'm doing wrong?

I even changed
```
    @Self
    @Via(type = ForcedResourceType.class, value = "core/wcm/components/form/container/v2/container")
    private Container superTypeContainer;
```
to
```
    @Self
    @Via(type = ResourceSuperType.class)
    private Container superTypeContainer;
```
to better match the delegation pattern as it is exactly documented, but I still get the infinite recursion.

**Comment by HitmanInWis** [comment]
Also, unless I am doing something incorrect in implementing the proxy pattern, can you please reopen this issue?

**Comment by jckautzmann** [comment]
@HitmanInWis I went ahead and created the following setup:
- `HS2Container` and `HS2ContainerImpl` with the code above
- a `hs2container` component with following super type:
`sling:resourceSuperType="core/wcm/components/form/container/v2/container"`
- in a page, a content node with a resource type pointing to the `hs2container` component
I requested the page and I didn't see any error in `error.log`.
I tested with Core Components 2.3.0 and a latest CQ build (6.5).

Can you please try the above setup and see if it works?

**Comment by HitmanInWis** [comment]
Not sure if this is still happening, but will reopen if we run into it again in the future.

**Comment by HitmanInWis** [comment]
I just tested this again on AEM 6.4 with WCM Core 2.5.0, and it still happens.

As long as my HTL file references the WCM Core `Container` interface there will be an infinite loop trying to instantiate my proxy impl class.

The source of the infinite loop appears to be this:
```
	at com.day.cq.wcm.foundation.forms.FormsHelper.includeResource(FormsHelper.java:598)
	at com.day.cq.wcm.foundation.forms.FormsHelper.runAction(FormsHelper.java:267)
	at com.adobe.cq.wcm.core.components.internal.models.v1.form.ContainerImpl.runActionTypeInit(ContainerImpl.java:146)
	at com.adobe.cq.wcm.core.components.internal.models.v1.form.ContainerImpl.initModel(ContainerImpl.java:136)```

When `superTypeContainer` is injected, it triggers the post construct of the WCM Core `ContainerImpl` class, which triggers an internal request include to the original resource. This gets adapted to my proxy model class as a new instance, which injects a new `superTypeContainer` instance, and the cycle repeats infinitely.

We're using proxy components for extending all other WCM Core components w/o issue, so it appears to apply to just this component.  Why you were not able to reproduce @jckautzmann I dont know but I'd be happy to take a look unless someone can tell me what I'm doing wrong in the code posted here.

**Comment by henrykuijpers** [comment]
@jckautzmann , can you post the code that successfully instantiates the subtype of form container? I'm very curious what is different in your code. Or did you also implement a workaround for this issue?

**Comment by jckautzmann** [comment]
@henrykuijpers Since I had removed the branch I used to reproduce this issue, I tried again today and recreated one. I tested with CQ 6.5 and Core Components 2.6.0 and was still not able to reproduce this issue. Here is the code I used:
https://github.com/adobe/aem-core-wcm-components/commit/14d538c8e181702756b53ab80b06229b29dfc5d2

**Comment by HitmanInWis** [comment]
Thanks @jckautzmann when I get a moment I'll try your code and see what's different between what I have and what you have.  Much appreciated.

**Comment by HitmanInWis** [comment]
@jckautzmann I tried your code and it didn't throw the infinite loop exception at first until I made sure the `resourceType` specified in the Impl model declaration was correct.

In your Impl you have `resourceType = "hs2/components/content/form/container"`

You would want to double check that:
- This resource type exists in the JCR
- Your form references this resource type
- The path is correct - note that the path more like WCM core would be "hs2/components/form/container" if you put it in a path similar to WCM Core

At first when I used your code it was not hitting the infinite loop, but that was b/c the Impl class was not actually being found in the first place due to `resourceType` not matching up.  Not saying this is exactly what you're hitting for sure, but seems possible especially b/c of the unintended "/content" slug in your resource type.

**Comment by ashish89** [comment]
I am also facing the same issue while extending the form container component. I am working on AEM 6.5 with core component 2.5.0. Is this issue resolved in new releases of core component ?

**Comment by HitmanInWis** [comment]
@ashish89 if you need a workaround, what I did was this in my sling model:

```
    private com.adobe.cq.wcm.core.components.models.form.Container superTypeContainer;

    @PostConstruct
    public void initContainerImplModel() {
        // This non-standard injection for the proxy model is due to an infinite loop triggered by
        // the standard proxy extension pattern.  See https://github.com/adobe/aem-core-wcm-components/issues/239.
        // Also note that for this to work I also needed to
        //   1) Remove `com.adobe.cq.wcm.core.components.models.form.Container` as an adapter from this class
        //   2) Update container.html to reference `com.bounteous.aem.base.core.models.form.Container`
        superTypeContainer = modelFactory.createModel(request, com.adobe.cq.wcm.core.components.models.form.Container.class);
    }
```

I then also removed `com.adobe.cq.wcm.core.components.models.form.Container.class` from the list of `adapters` in the `@Model` declaration, and updated `container.html` to reference my own `Container` interface (since I already was overriding `container.html` anyway this wasnt a big deal).

**Comment by ashish89** [comment]
Thanks @HitmanInWis I will try the solution you mentioned above.

**Comment by ashish89** [comment]
@HitmanInWis Its not working for me. I am doing like this:

`@Model(adaptables = SlingHttpServletRequest.class,
    adapters = {ContainerExtended.class},
    resourceType = {ContainerImpl.RESOURCE_TYPE_V1})
public class ContainerImpl implements ContainerExtended {`

` @PostConstruct
  private void initModel() {
    // This non-standard injection for the proxy model is due to an infinite loop triggered by
    // the standard proxy extension pattern.  See https://github.com/adobe/aem-core-wcm-components/issues/239.
    // Also note that for this to work I also needed to
    //   1) Remove `com.adobe.cq.wcm.core.components.models.form.Container` as an adapter from this class
    //   2) Update container.html to reference `com.bounteous.aem.base.core.models.form.Container`
    container = modelFactory.createModel(request, com.adobe.cq.wcm.core.components.models.form.Container.class);
  }`

Interface:
`@ConsumerType
public interface ContainerExtended extends Container {
  String getCustomAction();
}`

Container in interface is core component form container.

Please let me know if I am missing something.

**Comment by HitmanInWis** [comment]
Your `resourceType` in the `@Model` declaration should be your custom resource type, not the WCM Core type.

Also make sure your `container.html` is referencing `ContainerExtended` as the model class.

**Comment by ashish89** [comment]
@HitmanInWis Yes it is custom resource Type and also container.html is referencing ContainerExtended.

**Comment by HitmanInWis** [comment]
Can you post the full contents of your java impl here?  You removed the `@Self` and `@Via` annotations from the proxied container reference, correct?

**Comment by ashish89** [comment]
@HitmanInWis Yes it's working now. Thanks.

**Comment by vladbailescu** [comment]
I was able to reproduce it when adding the form container on a new page at top level. Added a possible fix at https://github.com/adobe/aem-core-wcm-components/commit/1c904b727cf50fad99dfa5b0414099225db400f2?w=1

**Comment by jckautzmann** [comment]
Thx @vladbailescu for trying out and providing a fix: after adding the custom form container to a new page, I was also able to reproduce it. I will commit Vlad's fix to the `development` branch.

**Comment by jckautzmann** [comment]
Fix provided in https://github.com/adobe/aem-core-wcm-components/pull/832

**Comment by andreeadracea** [comment]
The issue was fixed.

**Comment by richardhand** [comment]
The code for this issue is merged to master and will be available in the upcoming [2.8.0 release](https://github.com/adobe/aem-core-wcm-components/milestone/18).
